### PR TITLE
The commutative ring of rational numbers

### PR DESCRIPTION
**Pull request by malarbol**
This pull request introduces the commutative ring structure on the rational numbers.

We introduce:
- left distributivity of multiplication over the addition of integer fractions;
- (left|right) distributivity of multiplication over the addition of rational numbers (the last missing law for the ring structure);
- (left|right) distributivity of multiplication over the difference of rational numbers;
- a few laws about multiplication by zero in the rational numbers.

### CODE REVIEW

**Comment by malarbol** [comment]
I'm back for more `elementary-number-theory`. Hopefully this new PR will be less work than the last one. It's quite small and the changes are pretty straightforward.

**Comment by fredrik-bakke** [comment]
Do you know why I usually write `(left|right)` by the way? It's because that's how you write the regex to match either word :)

File: src/elementary-number-theory/multiplication-rational-numbers.lagda.md [hunk 1]
```diff
@@ -44,12 +45,70 @@ rational numbers.
...
+        ( ap (mul-ℤ' one-ℤ) (right-zero-law-mul-ℤ (numerator-ℚ x)))
+        ( left-zero-law-mul-ℤ _))) ∙
+    ( is-retraction-rational-fraction-ℚ zero-ℚ)
+```
+
+### If the product of two rational numbers is zero, the left or right factor is zero
+
+```agda
+is-zero-is-zero-mul-ℚ :
+  (x y : ℚ) → is-zero-ℚ (x *ℚ y) → (is-zero-ℚ x) + (is-zero-ℚ y)
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/multiplication-rational-numbers.lagda.md:89]: Since this is a decision procedure, maybe a clearer name is `decide-is-zero-factor-is-zero-mul-Q`?

File: src/elementary-number-theory/multiplication-rational-numbers.lagda.md [hunk 2]
```diff
@@ -179,4 +238,78 @@ negative-law-mul-ℚ x y =
...
+swap-neg-mul-ℚ x y =
+  ( right-negative-law-mul-ℚ x y) ∙
+  ( inv (left-negative-law-mul-ℚ x y))
+```
+
+### Multiplication on rational numbers distributes over addition
+
+```agda
+left-distributive-mul-add-ℚ :
+  (x y z : ℚ) → x *ℚ (y +ℚ z) ＝ x *ℚ y +ℚ x *ℚ z
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/multiplication-rational-numbers.lagda.md:252]: It seems to me like the right-hand side would read more clearly if the products were parenthesized

File: src/elementary-number-theory/ring-of-rational-numbers.lagda.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+</details>
+
+## Idea
+
+The
+[commutative group of rational numbers](elementary-number-theory.group-of-rational-numbers.md)
+equipped with
+[multiplication](elementary-number-theory.multiplication-rational-numbers.md) is
+a [commutative](commutative-algebra.commutative-rings.md)
+[ring](ring-theory.rings.md).
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/ring-of-rational-numbers.lagda.md:36]: I'm not sure this way of splitting the link is good practice. Its probably better if the composite term links to commutative ring. Note that if the reader wants to find the file on rings, they likely expect to find a link to it in the file on commutative rings.

File: src/elementary-number-theory/ring-of-rational-numbers.lagda.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+```agda
+ℚ-Ring : Ring lzero
+pr1 ℚ-Ring = ℚ-Ab
+pr1 (pr1 (pr2 ℚ-Ring)) = mul-ℚ
+pr2 (pr1 (pr2 ℚ-Ring)) = associative-mul-ℚ
+pr1 (pr1 (pr2 (pr2 ℚ-Ring))) = one-ℚ
+pr1 (pr2 (pr1 (pr2 (pr2 ℚ-Ring)))) = left-unit-law-mul-ℚ
+pr2 (pr2 (pr1 (pr2 (pr2 ℚ-Ring)))) = right-unit-law-mul-ℚ
+pr1 (pr2 (pr2 (pr2 ℚ-Ring))) = left-distributive-mul-add-ℚ
+pr2 (pr2 (pr2 (pr2 ℚ-Ring))) = right-distributive-mul-add-ℚ
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/ring-of-rational-numbers.lagda.md:51]: This definition is pretty long winded. In these cases, we prefer to make multiple smaller definitions. E.g. (although I don't recall the precise way we define rings)
define `is-monoid-mul-Q`, then `is-ring-Q` and then use those in your definition of the ring.

**Review by fredrik-bakke — APPROVED** [review]
Excellent contribution! I don't have much to comment on.

Note that I reviewed this from my phone, so formatting errors are hard to catch

File: src/elementary-number-theory/ring-of-rational-numbers.lagda.md (hunk 4 above)
Comment (fredrik-bakke) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:51]: It doesn't hurt to define all of the smaller parts you can

Comment (fredrik-bakke) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:51]: That constitute a part of the definition is `Ring` that is

File: src/elementary-number-theory/multiplication-rational-numbers.lagda.md (hunk 2 above)
Comment (fredrik-bakke) [reply src/elementary-number-theory/multiplication-rational-numbers.lagda.md:252]: And same for right distributive

File: src/elementary-number-theory/ring-of-rational-numbers.lagda.md (hunk 3 above)
Comment (malarbol) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:36]: This will soon become "a commutative division ring" and `[commutative division ring]` would redirect to [`discrete-fields`](https://unimath.github.io/agda-unimath/commutative-algebra.discrete-fields.html), which seems  a bit circular (there _will_ be a `discrete-field-of-rational-numbers` module importing `ring-of-rational-numbers` proving `is-discrete-field-Commutative-Ring ℚ-Commutative-Ring`). 

So it will have to be, at least `[commutative](xx) [division ring](xx)`, but I'd prefer `[commutative](commutative-algebra.commutative-rings.md) [division](ring-theory.division-rings.md) [ring](ring-theory.rings.md)` (one file per concept / one concept per file, etc.).

File: src/elementary-number-theory/ring-of-rational-numbers.lagda.md (hunk 4 above)
Comment (malarbol) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:51]: You know I just copy/pasted the [elementary-number-theory.ring-of-integers](https://unimath.github.io/agda-unimath/elementary-number-theory.ring-of-integers.html) module, right 😅 ?

```agda
ℤ-Ring : Ring lzero
pr1 ℤ-Ring = ℤ-Ab
pr1 (pr1 (pr2 ℤ-Ring)) = mul-ℤ
pr2 (pr1 (pr2 ℤ-Ring)) = associative-mul-ℤ
pr1 (pr1 (pr2 (pr2 ℤ-Ring))) = one-ℤ
pr1 (pr2 (pr1 (pr2 (pr2 ℤ-Ring)))) = left-unit-law-mul-ℤ
pr2 (pr2 (pr1 (pr2 (pr2 ℤ-Ring)))) = right-unit-law-mul-ℤ
pr1 (pr2 (pr2 (pr2 ℤ-Ring))) = left-distributive-mul-add-ℤ
pr2 (pr2 (pr2 (pr2 ℤ-Ring))) = right-distributive-mul-add-ℤ
```

Comment (malarbol) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md]: see https://github.com/UniMath/agda-unimath/pull/1107/commits/18d3e4046c29b960b2df494c981ffcf51bd7846c

Comment (fredrik-bakke) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md]: Our library is not perfect 🥲
> see https://github.com/UniMath/agda-unimath/pull/1107/commits/18d3e4046c29b960b2df494c981ffcf51bd7846c

Perfect!

File: src/elementary-number-theory/ring-of-rational-numbers.lagda.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+```
+
+## Definitions
+
+### The commutative ring of rational numbers
+
+```agda
+ℚ-Ring : Ring lzero
+pr1 ℚ-Ring = ℚ-Ab
+pr2 ℚ-Ring = has-mul-Ab-ℚ
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/ring-of-rational-numbers.lagda.md:69]: This is how we usually arrange this:
````suggestion
## Definitions

### The compatible multiplicative structure on the abelian group of rational numbers

```agda
has-associative-mul-ℚ : has-associative-mul ℚ
pr1 has-associative-mul-ℚ = mul-ℚ
pr2 has-associative-mul-ℚ = associative-mul-ℚ

is-unital-mul-ℚ : is-unital mul-ℚ
pr1 is-unital-mul-ℚ = one-ℚ
pr1 (pr2 is-unital-mul-ℚ) = left-unit-law-mul-ℚ
pr2 (pr2 is-unital-mul-ℚ) = right-unit-law-mul-ℚ

has-mul-ℚ-Ab : has-mul-Ab ℚ-Ab
pr1 has-mul-ℚ-Ab = has-associative-mul-ℚ
pr1 (pr2 has-mul-ℚ-Ab) = is-unital-mul-ℚ
pr1 (pr2 (pr2 has-mul-ℚ-Ab)) = left-distributive-mul-add-ℚ
pr2 (pr2 (pr2 has-mul-ℚ-Ab)) = right-distributive-mul-add-ℚ
```

### The ring of rational numbers

```agda
ℚ-Ring : Ring lzero
pr1 ℚ-Ring = ℚ-Ab
pr2 ℚ-Ring = has-mul-ℚ-Ab
```

### The commutative ring of rational numbers
```agda
````

Comment (fredrik-bakke) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:69]: note also that `has-mul-Ab-ℚ` should be named `has-mul-ℚ-Ab`

Comment (fredrik-bakke) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:69]: oh, whoops, this suggestion introduces a small formatting mistake, but it will be fixed by pre-commit

Comment (fredrik-bakke) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:69]: You know, this would be even better if we had a separate definition for distributivity of the multiplication operation. In fact, I am tempted to add a file about multiplication operations on abelian* groups.

File: src/elementary-number-theory/multiplication-rational-numbers.lagda.md [hunk 6]
```diff
@@ -44,12 +45,70 @@ rational numbers.
...
+
+  right-zero-law-mul-ℚ : x *ℚ zero-ℚ ＝ zero-ℚ
+  right-zero-law-mul-ℚ =
+    ( eq-ℚ-sim-fraction-ℤ
+      ( mul-fraction-ℤ (fraction-ℚ x) (fraction-ℚ zero-ℚ))
+      ( fraction-ℚ zero-ℚ)
+      ( eq-is-zero-ℤ
+        ( ap (mul-ℤ' one-ℤ) (right-zero-law-mul-ℤ (numerator-ℚ x)))
+        ( left-zero-law-mul-ℤ _))) ∙
+    ( is-retraction-rational-fraction-ℚ zero-ℚ)
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/multiplication-rational-numbers.lagda.md:82]: these should be abstracted

File: src/elementary-number-theory/multiplication-rational-numbers.lagda.md [hunk 7]
```diff
@@ -179,4 +238,78 @@ negative-law-mul-ℚ x y =
...
+            ( sim-reduced-fraction-ℤ
+              (add-fraction-ℤ (fraction-ℚ y) (fraction-ℚ z))))
+          ( refl-sim-fraction-ℤ (fraction-ℚ (y +ℚ z))))))
+
+right-distributive-mul-add-ℚ :
+  (x y z : ℚ) → (x +ℚ y) *ℚ z ＝ (x *ℚ z) +ℚ (y *ℚ z)
+right-distributive-mul-add-ℚ x y z =
+  ( commutative-mul-ℚ (x +ℚ y) z) ∙
+  ( left-distributive-mul-add-ℚ z x y) ∙
+  ( ap-add-ℚ (commutative-mul-ℚ z x) (commutative-mul-ℚ z y))
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/multiplication-rational-numbers.lagda.md:298]: abstract

File: src/elementary-number-theory/multiplication-rational-numbers.lagda.md [hunk 8]
```diff
@@ -179,4 +238,78 @@ negative-law-mul-ℚ x y =
...
+  (z x y : ℚ) → z *ℚ (x -ℚ y) ＝ (z *ℚ x) -ℚ (z *ℚ y)
+left-distributive-mul-diff-ℚ z x y =
+  ( left-distributive-mul-add-ℚ z x (neg-ℚ y)) ∙
+  ( ap ((z *ℚ x) +ℚ_) (right-negative-law-mul-ℚ z y))
+
+right-distributive-mul-diff-ℚ :
+  (x y z : ℚ) → (x -ℚ y) *ℚ z ＝ (x *ℚ z) -ℚ (y *ℚ z)
+right-distributive-mul-diff-ℚ x y z =
+  ( right-distributive-mul-add-ℚ x (neg-ℚ y) z) ∙
+  ( ap ((x *ℚ z) +ℚ_) (left-negative-law-mul-ℚ y z))
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/multiplication-rational-numbers.lagda.md:314]: abstract

**Comment by VojtechStep** [comment]
There's a malformed link in `elementary-number-theory.addition-rational-numbers` to `group-of-rational-numbers` that we didn't catch in the last PR - @malarbol would you mind fixing it in this PR?

File: src/elementary-number-theory/ring-of-rational-numbers.lagda.md (hunk 3 above)
Comment (fredrik-bakke) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:39]: I still believe it is unwanted to have two words in a composite term link to separate pages. It will not be clear to a reader that clicking on the word "ring" in the term "commutative ring" sends them to a different page than clicking on "commutative". Instead, if you want to have links to both, you should rephrase such that "ring" appears as a term by itself too. If this term is already important to the discussion then this will improve the prose too.

Comment (malarbol) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md:39]: ok. Let's change it and re-think about it later when we have to rephrase this to mention the division-ring property.

File: src/elementary-number-theory/ring-of-rational-numbers.lagda.md (hunk 4 above)
Comment (fredrik-bakke) [reply src/elementary-number-theory/ring-of-rational-numbers.lagda.md]: it can wait

**Comment by fredrik-bakke** [comment]
Oh, did you implement all the requested changes, @malarbol? Is this ready to merge?

**Comment by malarbol** [comment]
> Oh, did you implement all the requested changes, @malarbol? Is this ready to merge?

I think I did (have a last check maybe?) and that it's ready to merge.

**Review by fredrik-bakke — APPROVED** [review]
Looks great!

**Comment by fredrik-bakke** [comment]
I'll merge this right after #1096.
